api/python/quilt3/formats.py

```diff
@@ -1024,6 +1027,75 @@ def deserialize(self, bytes_obj, meta=None, ext=None, **format_opts):
 ParquetFormatHandler().register()  # latest is preferred
 
 
+# noinspection PyPackageRequirements
+class AnnDataFormatHandler(BaseFormatHandler):
+    """Format for AnnData <--> .h5ad
+
+    Format Opts:
+        The following options may be used anywhere format opts are accepted,
+        or directly in metadata under `{'format': {'opts': {...: ...}}}`.
+
+        compression('gzip', 'lzf', None):  applies during serialization only.
+    """
+    name = 'h5ad'
+    handled_extensions = ['h5ad']
+    opts = ('compression',)
+    defaults = dict(
+        compression='lzf',
+    )
+
+    def handles_type(self, typ: type) -> bool:
+        # don't load module unless we actually have to use it.
+        if 'anndata' not in sys.modules:  # pragma: no cover
+            return False
+        import anndata as ad
+        self.handled_types.add(ad.AnnData)
+        return super().handles_type(typ)
+
+    def serialize(self, obj, meta=None, ext=None, **format_opts):
+        import h5py
+
+        opts = self.get_opts(meta, format_opts)
+        opts_with_defaults = copy.deepcopy(self.defaults)
+        opts_with_defaults.update(opts)
+        try:
+            from anndata.experimental import write_elem
+        except ImportError:  # pragma: no cover
+            warnings.warn("anndata.experimenta.write_elem got moved", FutureWarning)
+            with TemporaryDirectory() as td:
+                path = Path(td) / 'data.h5ad'
+                obj.write(path, **opts_with_defaults)
+                data = path.read_bytes()
+        else:
+            buf = io.BytesIO()
+            with h5py.File(buf, "w") as f:
+                write_elem(f, "/", obj)
+            data = buf.getvalue()
+
+        return data, self._update_meta(meta, additions=opts_with_defaults)
+
+    def deserialize(self, bytes_obj, meta=None, ext=None, **format_opts):
+        try:
+            import anndata as ad
+        except ImportError:  # pragma: no cover
+            raise QuiltException("Please install anndata")
+
+        import h5py
```

Comment:
I think once we import `h5py`, it has to be declared as dependency of `quilt3[anndata]`.

Reply:
anndata directly depends on h5py, and it won’t stop doing so, so this will always be a transient dependency through anndata.

If you have a static dependency analyzer check that makes sure no dependency is undeclared, it’s a different story!